lib/segment/src/id_tracker/immutable_id_tracker.rs

```diff
@@ -0,0 +1,429 @@
+    }
+
+    fn external_id(&self, internal_id: PointOffsetType) -> Option<PointIdType> {
+        if *self.deleted.get(internal_id as usize)? {
+            return None;
+        }
+
+        self.mappings
+            .internal_to_external
+            .get(internal_id as usize)
+            .map(|i| i.into())
+    }
+
+    fn set_link(
+        &mut self,
+        _external_id: PointIdType,
+        _internal_id: PointOffsetType,
+    ) -> OperationResult<()> {
+        panic!("Trying to call a mutating function (`set_link`) of an immutable id tracker");
+    }
+
+    fn drop(&mut self, external_id: PointIdType) -> OperationResult<()> {
+        let internal_id = self
+            .mappings
+            .external_to_internal
+            .get(&external_id.into())
+            .copied();
+
+        if let Some(internal_id) = internal_id {
+            self.deleted.set(internal_id as usize, true);
+            *self.dirty.write() = true;
+        }
+
+        Ok(())
+    }
+
+    fn iter_external(&self) -> Box<dyn Iterator<Item = PointIdType> + '_> {
+        let iter_num = self
+            .mappings
+            .external_to_internal
+            .keys()
+            .filter(|i| i.is_num_id())
+            .map(|i| i.into());
+
+        let iter_uuid = self
+            .mappings
+            .external_to_internal
+            .keys()
+            .filter(|i| i.is_uuid())
+            .map(|i| i.into());
+        // order is important here, we want to iterate over the u64 ids first
+        Box::new(iter_num.chain(iter_uuid))
+    }
+
+    fn iter_internal(&self) -> Box<dyn Iterator<Item = PointOffsetType> + '_> {
+        Box::new(
+            (0..self.mappings.internal_to_external.len() as PointOffsetType)
+                .filter(move |i| !self.deleted[*i as usize]),
+        )
+    }
+
+    fn iter_from(
+        &self,
+        external_id: Option<PointIdType>,
+    ) -> Box<dyn Iterator<Item = (PointIdType, PointOffsetType)> + '_> {
+        let full_num_iter = || {
+            self.mappings
+                .external_to_internal
+                .iter()
+                .filter(|i| i.0.is_num_id())
+                .map(|(k, v)| (k.into(), *v))
+        };
+        let offset_num_iter = |offset: u64| {
+            self.mappings
+                .external_to_internal
+                .range(StoredPointId::NumId(offset)..)
+                .filter(|i| i.0.is_num_id())
+                .map(|(k, v)| (k.into(), *v))
+        };
+        let full_uuid_iter = || {
+            self.mappings
+                .external_to_internal
+                .iter()
+                .filter(|i| i.0.is_uuid())
+                .map(|(k, v)| (k.into(), *v))
+        };
+        let offset_uuid_iter = |offset: Uuid| {
+            self.mappings
+                .external_to_internal
+                .range(StoredPointId::Uuid(offset)..)
+                .filter(|i| i.0.is_uuid())
+                .map(|(k, v)| (k.into(), *v))
+        };
```

Comment:
Two separate types for numeric and UUID IDs are probably better because this may potentially be inefficient. We now have to go over the full list twice, with a more expensive iterator.

Reply:
Fixed in the other PR